### PR TITLE
support for x-www-form-urlencoded in access_tokens, refresh_tokens, revoke_tokens

### PR DESCRIPTION
**Pull request by Tekkon**
This addresses https://github.com/hintmedia/zoom_rb/issues/401 .

Our internal KUDO identifier is KMP-1946 .

Zoom has announced some security improvements to their API. On May 15th, 2022, their API will require that parameters in requests for access tokens, refresh tokens, and revoking tokens are included in the body of the POST instead of as URL parameters. This gem needs to be updated so that tokens and codes are part of the POST body instead of URL parameters for the access_tokens, refresh_tokens, and revoke_tokens calls.

This PR updates access_tokens, refresh_tokens, and revoke_tokens actions.

### CODE REVIEW

File: lib/zoom/client.rb [hunk 1]
```diff
@@ -29,20 +29,32 @@ class Client
     def headers
       {
         'Accept' => 'application/json',
-        'Content-Type' => 'application/json',
+        'Content-Type' => 'application/json'
```
Comment (gerryster) [inline lib/zoom/client.rb]: I would leave this comma to keep the diff churn down. As an aside, I am in favor of trailing commas to keep diff churn down when adding another element to these types of collections.

File: lib/zoom/utils.rb [hunk 2]
```diff
@@ -34,6 +34,13 @@ def parse_response(http_response)
       def extract_options!(array)
         params = array.last.is_a?(::Hash) ? array.pop : {}
         process_datetime_params!(params)
+        convert_param_names!(params)
+      end
+
+      def convert_param_names!(params)
```
Comment (gerryster) [inline lib/zoom/utils.rb]: This creates tight coupling between this module and `lib/zoom/actions/token.rb`. While I like these new names better as they match the API parameters exactly, I don't think that this is the right place to put this backward compatibility logic.

File: spec/lib/zoom/actions/token/revoke_token_spec.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  describe '#revoke_tokens action' do
+    before :each do
+      stub_request(
+        :post,
+        zoom_auth_url('oauth/revoke')
+      ).to_return(body: json_response('token', 'revoke_token'),
+                    headers: { 'Content-Type' => 'application/json' })
+    end
+
+    it "requires an error when args missing" do
```
Comment (gerryster) [inline spec/lib/zoom/actions/token/revoke_token_spec.rb]:
```suggestion
    it 'raises an error when args missing' do
```

File: lib/zoom/actions.rb [hunk 4]
```diff
@@ -40,10 +40,15 @@ def self.make_request(client, method, parsed_path, params, base_uri)
...
-          response = Zoom::Actions.make_request(self, method, parsed_path, params, base_uri)
+          response = Zoom::Actions.make_request(self, method, parsed_path, params, base_uri,
+                                                url_encoded: Zoom::Actions.form_url_encoded_actions.include?(name))
           Utils.parse_response(response)
         end
       end
     end
+
+    def self.form_url_encoded_actions
+      %w[access_tokens refresh_tokens revoke_tokens]
```
Comment (gerryster) [inline lib/zoom/actions.rb]: It seems odd that this generic meta-programming class which provides a DSL for defining actions also would know which specific actions require form URL encoding. This seems like something that the `lib/zoom/actions/token.rb` class should know about.

File: lib/zoom/client.rb [hunk 5]
```diff
@@ -29,20 +29,32 @@ class Client
...
       }
     end
 
+    def form_url_encoded_headers
+      headers.merge({ 'Content-Type' => 'application/x-www-form-urlencoded' })
```
Comment (gerryster) [inline lib/zoom/client.rb]: Minor: the curly braces aren't needed.

```suggestion
      headers.merge('Content-Type' => 'application/x-www-form-urlencoded')
```

**Review by gerryster — CHANGES_REQUESTED** [review]
This looks like a good start. My feedback is mostly about coupling and where to put the new logic. Also, I would like to see a test which verifies that the sensitive parameters are indeed part of the post body instead of URL parameters.

I also updated the description and title to emphasize the Github issue in the upstream Hintmedia repo and to downplay our internal identifiers which are meaningless in an open source context.

File: lib/zoom/actions.rb (hunk 4 above)
Comment (imstevecho) [reply lib/zoom/actions.rb]: I think @gerryster is right. Having this method in the token.rb makes more sense.

File: lib/zoom/actions.rb [hunk 6]
```diff
@@ -35,12 +35,15 @@ def self.make_request(client, method, parsed_path, params, base_uri)
 
         define_method(name) do |*args|
           path_keys = Zoom::Actions.extract_path_keys(path)
-          params = Zoom::Params.new(Utils.extract_options!(args))
+          params = Utils.extract_options!(args)
+          params = Zoom::Actions::Token.convert_param_names!(params)
```
Comment (gerryster) [inline lib/zoom/actions.rb]: This generic action module should now know about a specific action. One idea is that this module could define a `convert_param_names!` method which can be overridden by consuming classes such as the token action.

**Review by gerryster — CHANGES_REQUESTED** [review]
Looking better. I do have one last coupling concern. Also, taking a step back, does HTTPParty automatically use a `application/x-www-form-urlencoded` content type for post request with parameters?

**Review by imstevecho — APPROVED** [review]

**Comment by Tekkon** [comment]
@gerryster At Zoom::Actions.make_request headers are passed explicitly: { 'Accept' => 'application/json', 'Content-Type' => 'application/json' }, so it doesn't matter if HTTParty can use x-www-form-url-encoded automatically, we have to change the header in the request.

I changed the approach. I deleted convert_param_names! and form_url_encoded actions methods. I added additional params to Zoom::Actions :get, :post, :patch, :put, :delete methods. So, now it's possible to set what parameters should be renamed and additional headers in the post method call.

File: lib/zoom/actions.rb [hunk 7]
```diff
@@ -15,8 +15,8 @@ def self.parse_path(path, path_keys, params)
       parsed_path
     end
 
-    def self.make_request(client, method, parsed_path, params, base_uri, url_encoded: false)
-      request_options = { headers: url_encoded ? client.form_url_encoded_request_headers : client.request_headers }
+    def self.make_request(client, method, parsed_path, params, base_uri, headers = nil)
```
Comment (gerryster) [inline lib/zoom/actions.rb]: Five positional parameters is a bit much. I would suggest passing in a new object, a hash, or using keyword args.

File: lib/zoom/actions/token.rb [hunk 8]
```diff
@@ -8,12 +8,15 @@ module Token
       post 'access_tokens',
         '/oauth/token?grant_type=authorization_code',
         base_uri: 'https://zoom.us/',
-        require: %i[code redirect_uri]
+        require: %i[code redirect_uri],
+        args_to_params: { auth_code: :code },
+        headers: { 'Content-Type' => 'application/x-www-form-urlencoded' }
```
Comment (gerryster) [inline lib/zoom/actions/token.rb]: I would define `{ 'Content-Type' => 'application/x-www-form-urlencoded' }` in a constant such as `FORM_URLENCODED_HEADER`.

**Review by gerryster — APPROVED** [review]
LGTM with some minor suggestions. OK, the next step is to not merge this to the master branch of this repo but to instead create a PR to the master branch of https://github.com/hintmedia/zoom_rb and reference issue https://github.com/hintmedia/zoom_rb/issues/401 . Hopefully Hint Media will accept this change so we can rely on their main repo and use our fork as a staging area for these types of changes.

**Comment by Tekkon** [comment]
@gerryster I created the PR to hitmedia/zoom_rb repository: https://github.com/hintmedia/zoom_rb/pull/408. There were some changes in their repository, so I updated our fork and resolved the conflicts.

**Pull request merged**